src/local/room_name.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    error,
+    fmt::{self, Write},
+    ops,
+    str::FromStr,
+};
+
+use arrayvec::ArrayString;
+
+/// A structure representing a room name.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct LocalRoomName {
+    /// Inner x coordinate representation.
+    ///
+    /// 0 represents E0, positive numbers represent E(x)
+    ///
+    /// -1 represents W0, negative numbers represent W((-x) - 1)
+    pub x_coord: i32,
+    /// Inner y coordinate representation.
+    ///
+    /// 0 represents S0, positive numbers represent S(y)
+    ///
+    /// -1 represents N0, negative numbers represent N((-y) - 1)
+    pub y_coord: i32,
+}
+
+impl fmt::Display for LocalRoomName {
+    /// Formats this room name into the format the game expects.
+    ///
+    /// Resulting string will be `(E|W)[0-9]+(N|S)[0-9]+`, and will result
+    /// in the same LocalRoomName if passed into [`LocalRoomName::new`].
+    ///
+    /// [`LocalRoomName::new`]: struct.LocalRoomName.html#method.new
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.x_coord >= 0 {
+            write!(f, "E{}", self.x_coord)?;
+        } else {
+            write!(f, "W{}", (-self.x_coord) - 1)?;
+        }
+
+        if self.y_coord >= 0 {
+            write!(f, "S{}", self.y_coord)?;
+        } else {
+            write!(f, "N{}", (-self.y_coord) - 1)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl LocalRoomName {
+    /// Creates a new room name from the given input.
+    ///
+    /// This will parse the input, and return an error if it is in an invalid
+    /// format.
+    #[inline]
+    pub fn new<T>(x: &T) -> Result<Self, LocalRoomNameParseError>
+    where
+        T: AsRef<str> + ?Sized,
```

Comment:
What is this expecting? A `&str`? A `T: HasPosition`? Doc is unclear (and I honestly can't figure it out by just looking at the signature).

Reply:
Fair! I'll update the docs. It's expecting something that's string-like - `&str`, `&String`, `&ArrayString<[u8; N]>`, `Cow<'_, str>`, etc.

Reply:
Shouldn't it `Deref<&str>` or something?

Reply:
It could be, but I think `AsRef<&str>` is strictly more general?

For example, `Deref<&str>` is implemented for `String`, but not for `&String`.

Reply:
I'm blind: I noticed the `AsRef`, but not the `str`. It makes so much more sense now!